File: rasnet/src/service/server/servicerequesthandler.cc
/*
 * This file is part of rasdaman community.
 *
 * Rasdaman community is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Rasdaman community is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with rasdaman community.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Copyright 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014 Peter Baumann / rasdaman GmbH.
 *
 * For more information please see <http://www.rasdaman.org>
 * or contact Peter Baumann via <baumann@rasdaman.com>.
 */

#include <stdexcept>
#include <string>
#include <utility>

#include "../../../../common/src/logging/easylogging++.hh"

#include "../../../src/service/server/servicerequesthandler.hh"
#include "../../../src/messages/service.pb.h"
#include "../../../src/service/server/servercontroller.hh"
#include "../../../src/util/proto/protozmq.hh"

namespace rasnet
{
using std::runtime_error;
using rasnet::ServiceRequest;
using rasnet::ServiceResponse;
using std::string;
using std::pair;
using std::map;
using google::protobuf::Service;
using google::protobuf::MethodDescriptor;
using google::protobuf::Closure;
using google::protobuf::Message;
using google::protobuf::DoNothing;
using google::protobuf::NewCallback;
using boost::unique_lock;
using boost::shared_lock;
using boost::shared_mutex;
using boost::shared_ptr;


ServiceRequestHandler::ServiceRequestHandler(zmq::context_t& context,
        std::string& bridgeAddress) : context(context),bridgeAddress(bridgeAddress)
{
    this->messageType=ServiceRequest::default_instance().GetTypeName();
    this->doNothingClosure = google::protobuf::NewPermanentCallback(&DoNothing);
}

ServiceRequestHandler::~ServiceRequestHandler()
{
    delete this->doNothingClosure;
    this->serviceMethodMap.clear();
    this->serviceMap.clear();
}


void ServiceRequestHandler::addService(boost::shared_ptr<google::protobuf::Service> service)
{
    unique_lock<shared_mutex> lock(this->serviceMutex);
    //Get the full name of the service and use it as a key
    string serviceName = service->GetDescriptor()->full_name();
    pair<map<string, shared_ptr<google::protobuf::Service> >::iterator, bool> insertRes =
        this->serviceMap.insert(pair<string, shared_ptr<google::protobuf::Service> >(serviceName, service));
    if(insertRes.second==true)
    {
        //mark the ownership of the
        for (int i = 0; i < service->GetDescriptor()->method_count(); i++)
        {
            const MethodDescriptor* method =
                service->GetDescriptor()->method(i);
            //Use the full name of the method as a key
            string methodName = ProtoZmq::getServiceMethodName(method);
            this->serviceMethodMap[methodName] =method;
        }
    }
    else
    {
        throw runtime_error("Service already exists");
    }
}

bool ServiceRequestHandler::canHandle(const base::BaseMessage& message)
{
    return (message.type() == this->messageType);
}

void ServiceRequestHandler::handle(const base::BaseMessage& message,
                                   const std::string& peerId)
{
    ServerController controller;
    ServiceResponse response;
    shared_ptr<Service> service;

    if(this->canHandle(message))
    {
        ServiceRequest request;
        if(request.ParseFromString(message.data()))
        {
            map<string, const MethodDescriptor*>::iterator it;
            //Set the it for the response
            response.set_id(request.id());

            shared_lock<shared_mutex> lock(this->serviceMutex);

            it = this->serviceMethodMap.find(request.method_name());
            if(it!= this->serviceMethodMap.end())
            {
                map<string, shared_ptr<Service> >::iterator serviceIterator;
                const MethodDescriptor* methodDesc = it->second;
                serviceIterator=this->serviceMap.find(methodDesc->service()->full_name());

                if(serviceIterator!=this->serviceMap.end())
                {
                    service=serviceIterator->second;
                    //From this point on, we don't need exclusive access to the
                    //private datastructures so unlock the mutex

                    lock.unlock();

                    //Create objects to hold the request and response data
                    Message* replyData = service->GetResponsePrototype(
                                             methodDesc).New();
                    Message* requestData = service->GetRequestPrototype(
                                               methodDesc).New();

                    //Deserialize the input value
                    if (requestData->ParseFromString(request.input_value()))
                    {
                        //call the method
                        try
                        {
                            service->CallMethod(methodDesc, &controller,
                                                requestData, replyData, this->doNothingClosure);
                        }
                        catch(std::exception& ex)
                        {
                            LERROR<<ex.what();
                            response.set_error(ex.what());
                        }
                        catch(...)
                        {
                            LERROR<<"Service call to method" <<service->GetDescriptor()->full_name()<<" with input data" << requestData->SerializeAsString()<<" failed for unknown reason";
                            response.set_error("Function call failed on server");
                        }

                        if (controller.Failed())
                        {
                            response.set_error(controller.ErrorText());
                        }
                        else
                        {
                            response.set_output_value(
                                replyData->SerializeAsString());
                        }
                    }
                    else
                    {
                        LERROR<<"Invalid input message: "+message.DebugString();
                        response.set_error("Invalid input message: "+message.DebugString());
                    }

                    delete requestData;
                    delete replyData;
                }
                else
                {
                    response.set_error(
                        "This service is not offered by the server.");
                }
            }
            else
            {
                response.set_error(
                    "This method "+request.method_name() +" is not offered by the server.");
                LERROR<<"Call to inexisting method"<<request.method_name();
            }
        }
        else
        {
            //TODO:Document the fact that 0 is a special message ID.
            response.set_id(0);
            response.set_error("Could not parse message from client"+request.DebugString());
            LERROR<<"Could not parse message from client"<<request.DebugString();
        }

        //Finally, send the response
        this->sendMessageToBridge(response, peerId);
    }
    else
    { //Programmer error
        throw runtime_error("The handler cannot process this type of message:"+message.GetTypeName());
    }
}

void ServiceRequestHandler::sendMessageToBridge(
    google::protobuf::Message& mess, const std::string& peerId)
{
    zmq::socket_t peer(context, ZMQ_DEALER);
    peer.connect(bridgeAddress.c_str());
    ProtoZmq::zmqSendToPeer(peer, mess,peerId);
}
}

File: java/src/org/rasdaman/rasnet/server/ServiceRequestHandler.java
/*
 * This file is part of rasdaman community.
 *
 * Rasdaman community is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Rasdaman community is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with rasdaman community.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Copyright 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014 Peter Baumann / rasdaman GmbH.
 *
 * For more information please see <http://www.rasdaman.org>
 * or contact Peter Baumann via <baumann@rasdaman.com>.
 */

package org.rasdaman.rasnet.server;

import com.google.protobuf.*;
import org.rasdaman.rasnet.common.Constants;
import org.rasdaman.rasnet.exception.DuplicateService;
import org.rasdaman.rasnet.exception.UnsupportedMessageType;
import org.rasdaman.rasnet.message.Communication;
import org.rasdaman.rasnet.message.internal.Internal.ServiceResponseAvailable;
import org.rasdaman.rasnet.util.ZmqUtil;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.zeromq.ZMQ;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentLinkedQueue;

/**
 * @brief The ServiceRequestHandler class Handles service requests from
 * client Channels by calling the appropriate server implementation.
 */
public class ServiceRequestHandler {
    private static final Logger LOG = LoggerFactory.getLogger(ServiceRequestHandler.class);

    private ZMQ.Context context;
    /**
     * Address of the ROUTER socket in the ServiceManager worker thread that is responsible for
     * receiving messages from threads processing the service requests
     */
    private String bridgeAddress;

    /**
     * Map between the service's fully qualified name and the service
     */
    private ConcurrentHashMap<String, Service> serviceMap;

    /**
     *  Map between the fully qualified name of the method and the method descriptor
     */
    private ConcurrentHashMap<String, Descriptors.MethodDescriptor> serviceMethodMap;

    /**
     * Queue containing the results of completed service calls
     */
    private ConcurrentLinkedQueue<ServiceResponse> completedTasksResults;

    /**
     * @brief ServiceRequestHandler
     * @param context ZMQ context used by the ServerManager that owns this object
     * for internal communication
     * @param bridgeAddress Address of the bridge used to forward service responses
     * from this thread, to the worker thread in the ServerManager to the client
     */
    public ServiceRequestHandler(ZMQ.Context context, String bridgeAddress) {
        this.context = context;
        this.bridgeAddress = bridgeAddress;
        this.serviceMap = new ConcurrentHashMap<String, Service>();
        this.serviceMethodMap = new ConcurrentHashMap<String, Descriptors.MethodDescriptor>();
        this.completedTasksResults = new ConcurrentLinkedQueue<ServiceResponse>();
    }

    /**
     * Add the service to the list of available services.
     * @param service
     * @throws DuplicateService If the service is already in the list, an exception will be thrown.
     */
    public void addService(Service service) throws DuplicateService {
        if (service == null) {
            throw new IllegalArgumentException("service");
        }

        String serviceName = service.getDescriptorForType().getFullName();
        Service oldService = this.serviceMap.put(serviceName, service);

        if (oldService != null) {
            throw new DuplicateService();
        } else {
            List<Descriptors.MethodDescriptor> methods = service.getDescriptorForType().getMethods();
            Iterator<Descriptors.MethodDescriptor> it = methods.iterator();

            while (it.hasNext()) {
                Descriptors.MethodDescriptor method = it.next();
                String methodName = method.getFullName();
                serviceMethodMap.put(methodName, method);
            }
        }
    }

    /**
     * @brief getResponse Get the next available service response
     */
    ServiceResponse getTaskResult() {
        return this.completedTasksResults.poll();
    }

    /**
     * Check if the given message represents a service request
     * @param message   Message format :| MessageType = SERVICE_REQUEST | Call ID | Method Name | Serialized Input Data|
     * @return true if the message can be processed, false otherwise
     */
    public boolean canHandle(ArrayList<byte[]> message) {
        boolean success = false;

        if (message.size() == 4) {
            Communication.MessageType type;
            try {
                type = Communication.MessageType.parseFrom(message.get(0));
                success = (type.getType() == Communication.MessageType.Types.SERVICE_REQUEST);
            } catch (InvalidProtocolBufferException e) {
                e.printStackTrace();
                success = false;
            }
        }

        return success;
    }

    /**
     * Process the message if it contains a service request, throw an exception otherwise.
     * The exception is thrown only in case of programmer error i.e. Calling the method
     * on a message which has not been verified with canHandle
     * Call the appropriate method and return a result to the requesting client.
     * @param message BaseMessage containing a ServiceRequest
     * @param peerId ID of the peer that requested the service
     * and which can be used to reply to the service call.
     */
    public void handle(ArrayList<byte[]> message, String peerId) throws UnsupportedMessageType {
        //boolean which indicates if the callback was called.
        //If the callback was not called, an error is reported to the user.
        final boolean[] callbackCalled = {false};
        if (this.canHandle(message)) {
            //| MessageType = SERVICE_REQUEST | Call ID | Method Name | Serialized Input Data|
            String callId = new String(message.get(1));
            String methodName = new String(message.get(2));
            final ServiceResponse serviceResponse = new ServiceResponse(callId);

            final ServerController controller = new ServerController();
            Descriptors.MethodDescriptor methodDescriptor = this.serviceMethodMap.get(methodName);
            if (methodDescriptor != null) {
                //Retrieve the service associated with the method descriptor
                Service service = this.serviceMap.get(methodDescriptor.getService().getFullName());
                if (service != null) {
                    try {
                        final String replyDataType = ZmqUtil.getType(service.getResponsePrototype(methodDescriptor).newBuilderForType().buildPartial());

                        Message requestData = service.getRequestPrototype(methodDescriptor).getParserForType().parseFrom(message.get(3));

                        //Call the method and set the output
                        service.callMethod(methodDescriptor, controller, requestData, new RpcCallback<Message>() {
                            @Override
                            public void run(Message parameter) {
                                callbackCalled[0] = true;

                                //Sanity check
                                if (replyDataType.equals(ZmqUtil.getType(parameter))) {
                                    serviceResponse.setOutputValue(parameter.toByteString());
                                } else {
                                    //Invalid message received.
                                    controller.setFailed("The response set in the callback is invalid. Contact the service implementer.");
                                }
                            }
                        });

                        if (!callbackCalled[0] && !controller.failed()) {
                            controller.setFailed("Invalid method implementation. Contact the service implementer.");
                        }

                    } catch (InvalidProtocolBufferException e) {
                        LOG.error(e.getLocalizedMessage());
                        controller.setFailed("The input data is unparsable.");
                    } catch (Exception ex) {
                        LOG.error(ex.getLocalizedMessage());
                        controller.setFailed(ex.getLocalizedMessage());
                    }
                } else {
                    String errorMessage = "There is no service with the given name offered by this server.";
                    LOG.error(errorMessage);
                    controller.setFailed(errorMessage);
                }
            } else {
                String errorMessage = "There is no method with the given name on this server.";
                LOG.error(errorMessage);
                controller.setFailed(errorMessage);
            }

            if (controller.failed()) {
                serviceResponse.setError(controller.errorText());
            }

            this.completedTasksResults.add(serviceResponse);

            this.sendMessageToBridge(peerId);

        } else {
            throw new UnsupportedMessageType();
        }
    }

    /**
     * Create a ZeroMQ DEALER socket and connect it to the bridge_addr
     * and then send the message to that address
     * @param peerId ID of the peer that sent the message that is being handled.
     */
    private void sendMessageToBridge(String peerId) {
        try {
            ZMQ.Socket socket = this.context.socket(ZMQ.DEALER);
            socket.setLinger(0);
            socket.setIdentity(peerId.getBytes(Constants.DEFAULT_ENCODING));

            socket.connect(this.bridgeAddress);

            ServiceResponseAvailable responseAvailable = ServiceResponseAvailable.getDefaultInstance();
            ZmqUtil.send(socket, responseAvailable);

            socket.disconnect(this.bridgeAddress);
            socket.close();
        } catch (org.zeromq.ZMQException ex) {
            LOG.error(ex.getMessage());
        } catch (Exception ex) {
            LOG.error(ex.getMessage());
        }
    }
}
